File: javascript/anagram/anagram.js
const isAnagram = (a, b) => a !== b && [...a].sort().join('') === [...b].sort().join('');

export class Anagram {
    constructor(word) {
        this.word = word.toLowerCase();
    }

    matches(candidates) {
        return candidates.filter((c) => isAnagram(this.word, c.toLowerCase()));
    }
}

File: kotlin/anagram/src/main/kotlin/Anagram.kt
class Anagram(val input: String) {
    private val sortedInput = input.toLowerCase().toList().sortedBy { it }

    fun match(candidates: List<String>): Set<String> = if (input.toList().all { it.isUpperCase() }) {
        // This requirement is weird.
        candidates.intersect(listOf(input)).toSet()
    } else {
        candidates.filter { it.toLowerCase().toList().sortedBy { it }.equals(sortedInput) }.toSet()
    }
}
